Elaf/ulf surface tension

elafnm opened the pull request:
Greetings dears @playpaolo  and @pooyan-dadvand and very good day to you,

Reference to our very fruitful discussion on issue # 941,and  based on your appreciated recommendations that we received on Nov 3, 2017, We would like to inform you that we completed all the recommended steps, and the code is running successfully now for 2D Lagrangian case. Accordingly, we are raising the first pull request for 2D droplet model (Lagrangian and sessile droplets). 

Please note also that we added two test files called "Lagrangian_droplet_test" and "Lagrangian_sessile_droplet_test", with their results as well for further validation as needed. 

We appreciate your support and acceptance as  this will help us to move forward to the 3D models.

Thanks and best regards

elafnm committed:
some variable included

elafnm committed:
change to applications/CMakeList.txt

elafnm committed:
restored for compiling

elafnm committed:
adding SurfaceTension2D

elafnm committed:
adding SurfaceTension2D variables and elements

elafnm committed:
$.. added to Cmake.txt inside ULF application

elafnm committed:
added/modefied files for surface tension/lagrangian droplet model

elafnm committed:
changed

elafnm committed:
updated changes

elafnm committed:
updated changes

elafnm committed:
updated changes

elafnm committed:
updated changes

elafnm committed:
updated changes

elafnm committed:
updated changes

elafnm committed:
updated changes

elafnm committed:
updated

elafnm committed:
updated

elafnm committed:
updated

elafnm committed:
updated

elafnm committed:
updated

elafnm committed:
updated

elafnm committed:
updated

elafnm committed:
updated

pooyan-dadvand commented:
It seems that there are conflicts with @KratosMultiphysics/fluid-dynamics applications, so I added @jcotela as reviewer

jcotela commented on applications/ALEapplication/python_scripts/mesh_solver_base.py:
```diff
@@ -3,6 +3,9 @@
 import KratosMultiphysics
 import KratosMultiphysics.ALEApplication as KratosALE
 KratosMultiphysics.CheckForPreviousImport()
+import mesh_solver_base
```
Is this file importing itself?

philbucher replied:
It is, can you please revert the changes in this file?

playpaolo replied:
Note that you should not change the existing files of other applications (such as ALE one here) until unless it is completely necessary. Is mesh_solver_base a standard file (the existing one) or is it added by you?

elafnm replied:
It a standard file, 

sure will do dears,

jcotela commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -17,6 +17,14 @@
+#include <ctime>
+#include <stdlib.h>
+#include <iomanip>
+
```
Why are you adding all these includes? You don't seem to be using them in the new code.

jcotela commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -398,14 +419,24 @@ namespace Kratos {
-            // if (rModelPart.GetCommunicator().MyPID() == 0)
-            //     std::cout << "end of prediction" << std::endl;
+             if (rModelPart.GetCommunicator().MyPID() == 0)
+                 std::cout << "end of prediction" << std::endl;
```
These kinds of prints are fine for debug, but we would prefer not to have them in the final version. Please comment them out again or replace them by calls to the kratos logger.

elafnm replied:
@jcotela 

done dear, these two lines are commented now,

jcotela commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -328,8 +336,21 @@ namespace Kratos {
-                        UpdateDisplacement(CurrentDisplacement, OldDisplacement, OldVelocity, OldAcceleration, CurrentAcceleration);
+    
+                        
+                        if((itNode)->FastGetSolutionStepValue(IS_LAGRANGIAN_INLET) == 0.0)
```
I'd suggest using a flag instead of a nodal variable, it will save you memory.

elafnm replied:
Greetings dear @jcotela 

I hope I have done the required changes so far correctly, where I removed the (.directory), (.bin), and changed the header,

noting also please that I tried to follow your valuable recommendations for merging the master into this branch to fruther resolve all the conflicts, and I think it worked fine,, 

regarding using the flag instead the nodal variables, I agree with you, but kindly I will just work it out with Dr. Jarauta since he is the one who developed this code, and will include it soon.

Thanks and best regards,

elafnm replied:
Greetings dear @jcotela and very good day to you,

Kindly, we need your feedback regarding the flag: 

could you please elaborate more about the recommended way to be used within kratos, because from what I read in kratos forum website "http://kratos-wiki.cimne.upc.edu/forum/index.php?topic=36.0" that it is not recommended to add/define a local flag variable, and also I was thinking that it will not be convenient to link it with fluid dynamic applications.



Thanks and our best regards,

jcotela replied:
@elafnm: ideally, you would use one of the existing global flags (see post 4 in your link) instead of a nodal variable. If you don't think they are convenient for you, then just keep the current implementation. I also agree that adding a new local flag is not a good solution in this case.

elafnm replied:
Greetings dear @jcotela and very good day to you,

If you do not mind please, let us just keep the current implementation, this would be more convenient to us.

thank you for your usual support and very positive feedback,

jcotela commented on applications/FluidDynamicsApplication/fluid_dynamics_application.h:
```diff
@@ -33,12 +33,14 @@
 #include "includes/kratos_application.h"
 
+#include "includes/deprecated_variables.h"
+
```
Why are you including those here? I don't see them used on this file.

elafnm replied:
greetings dear @jcotela ,

so far, I am almost done with the modifications, but only one thing please to make sure I got you correctly,
do you want me to include these below elements, and also to include their related documents:

#include "custom_elements/qs_vms.h"
#include "custom_elements/symbolic_navier_stokes.h"
#include "custom_elements/embedded_fluid_element.h"

Thanks and regards,

jcotela replied:
Dear @elafnm: the elements you mention and the related code have been added to the fluid dynamics application while you were developing your code. Please do not remove them.

elafnm replied:
thank you very much dear @jcotela for your prompt reply, 

sure will do,

just would like kindly to know also if I need to add these elements in my branch as well and compile it in my machine to resolve the conflict, 

because when I tried to add these three lines only and try to compile kratos, it complained that these elements are not existing, or you think no need for adding them in my branch for this pull request!

thanks again and my best regards,

jcotela replied:
Dear @elafnm: since your changes to this file are very minor, I'd suggest just removing your modifications (the `#include "includes/deprecated_variables.h"` above). hopefully then the file can be merged without further conflicts.
If you ever find yourself in a similar situation, but want to keep the changes, I'd suggest merging the master to your branch (which will add all required files) and solving the conflict from there.

jcotela commented on applications/FluidDynamicsApplication/fluid_dynamics_application.h:
```diff
@@ -33,12 +33,14 @@
 //#include "custom_conditions/fluid_periodic_condition_2d.h"
 #include "custom_elements/vms.h"
-//#include "custom_elements/dynamic_vms.h"
+#include "custom_elements/dynamic_vms.h"
```
Again, this file is not using this element (and I should really remove it completely from the repository, but that's another story...).

jcotela reviewed (changes requested):
There are some issues that should be addressed before this PR is approved:
- The `.directory` files should not be in the repository, since they are specific to your system and may conflict with those of other users. Please remove them from the branch. In case you did not know, you can tell git to ignore selected local files so that they are never included in your commits. We have a small tutorial explaining this here: https://github.com/KratosMultiphysics/Kratos/wiki/How-to-make-Git-ignore-the-files-resulting-from-a-compilation-without-conflicts-in-.gitignore
- I'd also suggest not including binary gid output (`.bin` files) in the repository, since these files tend to be big and do not play well with git's version control. Please remove them from your examples. If you want, you could consider uploading those same test examples to the Kratos Examples repository instead. https://github.com/KratosMultiphysics/Examples
- Regarding the merge conflict in fluid_dynamics_application.h, if it is OK, I'd suggest just ignoring your changes and using the current version of the file. If some other file needs the deprecated variables, just add the include in that file.
- I've also included several comments and questions in the code itself.
- I'd suggest using the new version of the kratos license comments (you can use the comment at the top of this file here as a reference: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/templates/header_template).
- I do not have a lot to say about the ULF-specific stuff. Please wait for @playpaolo's comments on that part.

philbucher commented on applications/ALEapplication/python_scripts/mesh_solver_laplacian.py:
```diff
@@ -6,6 +6,16 @@
+
+
+
+def AddDofs(model_part):
```
This function does not belong there, it should be in your script

philbucher reviewed (changes requested):
2 minor comments

rubenzorrilla commented on applications/FluidDynamicsApplication/custom_elements/vms.h:
```diff
@@ -1895,4 +1895,4 @@ inline std::ostream& operator <<(std::ostream& rOStream,
 } // namespace Kratos.
 
-#endif // KRATOS_VMS_H_INCLUDED  defined
+#endif // KRATOS_VMS_H_INCLUDED  defined
```
Probably your code editor automatically did this change. Please do not add these kind modifications to your PRs, since they make the revision much more tedious.

elafnm replied:
sorry about that dear, 

sure will do and will take care in future,

rubenzorrilla commented on applications/MeshingApplication/custom_python/add_meshers_to_python.cpp:
```diff
@@ -146,6 +148,19 @@ void TriRegenerateMeshGLASS(TriGenGLASSModeler& Mesher, char* ElementName, char*
+//											//
+//////////////////////////////////////////////////////////////////////////////////////////
+
+void TriRegenerateMeshDROPLET(TriGenDropletModeler& Mesher, char* ElementName, char* ConditionName, ModelPart& model_part,NodeEraseProcess& node_erase, bool rem_nodes, bool add_nodes, double alpha_shape, double h_factor )
```
Are you sure that this function declaration must be placed in this file? Probably you can set up an utility or process and then just implement the Python exportation in this file.

playpaolo replied:
I believe that this is done fine.

rubenzorrilla commented on applications/ULFapplication/CMakeLists.txt:
```diff
@@ -21,9 +21,12 @@ set( KRATOS_ULF_APPLICATION_SOURCES
 	${CMAKE_CURRENT_SOURCE_DIR}/custom_conditions/Point_Neumann2D.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/custom_conditions/Point_Neumann3D.cpp
-	${CMAKE_CURRENT_SOURCE_DIR}/custom_conditions/Point_Neumann_Axisym.cpp
+ 	${CMAKE_CURRENT_SOURCE_DIR}/custom_conditions/Point_Neumann_Axisym.cpp
```
Follow the file naming convention (change Point _Neumann_Axisym.cpp to point_neumann_axisym.cpp)

playpaolo replied:
Have these changes been done by @elafnm ? Point_Neumann_Axisym.* are my elements. In any case, we should change the name following the convention.

rubenzorrilla replied:
Probably @elafnm 's code editor automatically removed the last character of the line and this is why the file appears as edited.

rubenzorrilla commented on applications/ULFapplication/Lagrangian-droplet_test/Lap_young_lagr.py:
```diff
@@ -0,0 +1,244 @@
+from __future__ import print_function, absolute_import, division #makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
Same naming convention comments (for both the folder and the file name).

rubenzorrilla commented on applications/ULFapplication/Lagrangian-droplet_test/ProjectParameters.py:
```diff
@@ -0,0 +1,89 @@
+domain_size = 2
```
I guess that your application still works with the old ProjectParameters Python scripts. About a year ago, we decided to use the .json file format for the ProjectParameters, meaning that the old ProjectParameters file format became deprecated. It would be a good idea to take the chance to update that.

playpaolo replied:
At the moment I disagree with this since it would require changing the whole ULF application. ULF is not working with the same problem type as the one people of FluidDynamicsApplication are using. Our problemtype write the input parameters files in the old format. We can discuss passing to the new format, but I believe this is not necessary for this pull.

rubenzorrilla commented on applications/ULFapplication/ULF_application.cpp:
```diff
@@ -19,31 +19,78 @@
+//     KRATOS_CREATE_VARIABLE(double, IS_LAGRANGIAN_INLET)
+    //KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(PRESSURE_FORCE)
+    //KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(DISP_FRAC)
+    KRATOS_CREATE_VARIABLE(double, PRESSURE_OLD_IT)
```
This variable already exists. Indeed, multiple variables in here already exist. As far as I can understand this might be a potential problem...

elafnm replied:
Greetings dear @rubenzorrilla,

I just check about the variables added in the ULF applications, and nothing is duplicated with the "variables.h" or "variables.cpp" inside "kratos/kratos/includes" or "kratos/kratos/sources". However, this "PRESSURE_OLD_IT" for example is defined inside the cfd_variables... and "PRESSURE_FORCE" is defined inside the meshing_applications,,,, so what do you recommend in this case please

rubenzorrilla replied:
You have two options in here, 

- Include the files that declare these variables.
- Or defining a new local variable in your application.

For the PRESSURE_OLD_IT, I would include the cfd_variables file since it belongs to the core. For the PRESSURE_FORCE, do whatever you want, but in any case, do not declare it again.

elafnm replied:
thanks dear, 

sure will do,

so I will go with the first option always,

elafnm replied:
Greetings dear @rubenzorrilla ,

I hope I have done the required as required, kindly note that I have I removed the variables that are already existing, and also I defined the needed variables fir this code locally inside ULF applications. Also, I tried to follow the naming conventions for the added files, with the recommended header format,

Thanks and best regards,

rubenzorrilla commented on applications/ULFapplication/custom_processes/find_nodal_h_process.h:
```diff
@@ -182,7 +182,7 @@ class FindNodalHProcess
         //calculate average h
         if(h_nodes == 0)
-            KRATOS_THROW_ERROR(std::logic_error,"no node has neighbours!!!!","");
+            KRATOS_ERROR(std::logic_error,"no node has neighbours!!!!","");
```
There's a common find_nodal_h_process in Kratos Core. It would be nice if you could use that one to avoid code duplication.

rubenzorrilla commented on applications/ULFapplication/custom_processes/find_nodal_neighbours_surface_process.h:
```diff
@@ -0,0 +1,382 @@
+/*
```
This is the old file header. Use the current one (you have already used it in the majority of your implementations).

rubenzorrilla commented:
I'm adding @AndreasWinterstein as reviewer since he is in charge of the ALE application. Thank you Andreas!

elafnm committed:
Delete .directory

playpaolo commented:
Apart from the particular comments made above I have overall suggestions:

1) avoid changing the "standard files" of other applications
2) make sure that the variables you define are not defined already elsewhere
3) do not put very specific variables into kratos core variabl files (vartiables.h and variables.cpp), but rather define it locally in your application (ULF in this case)
4) Make sure that the variable names are comprehensive
5) When adding new test_examples to the repository make sure that you add only the necessary files (no results (bins), no auxiliaries). Typically, these are 3: the *.mdpa file with the model, the parameters file and the example script.

elafnm commented:
thank you all for your comments and support, I am working on correcting them as recommended,

elafnm committed:
updated

elafnm committed:
Merge branch 'elaf/ULF_surface_tension' of https://github.com/KratosMultiphysics/Kratos into elaf/ULF_surface_tension

elafnm committed:
changed for pull request

elafnm committed:
variables.h updated

elafnm committed:
Merge branch 'master' into elaf/ULF_surface_tension

elafnm committed:
merging master to my  branch to resolve  conflicts

elafnm committed:
Delete .directory

elafnm committed:
Delete .directory

elafnm committed:
Delete .directory

elafnm committed:
Delete .directory

elafnm committed:
required changes

elafnm committed:
Merge branch 'elaf/ULF_surface_tension' of https://github.com/KratosMultiphysics/Kratos into elaf/ULF_surface_tension

elafnm commented:
Greetings dear @philbucher,

I hope I have done your recommendations successfully, where I revert the two ALE files mentioned by you. 

Thanks and best regards,

elafnm commented:
Greetings dears @playpaolo  and very good day to you,

Just would like to inform that I merged the master branch to this branch to resolve the conflicts, and also  tried my best to follow and fulfil your recommendations so far, 

thanks a lot and my best regards,

philbucher commented on applications/ALEapplication/python_scripts/mesh_solver_base.py:
```diff
@@ -127,4 +127,4 @@ def _create_mesh_motion_solver(self):
         The mesh motion solver must provide the functions defined in SolutionStrategy.
         """
-        raise Exception("Mesh motion solver must be created by the derived class.")
+        raise Exception("Mesh motion solver must be created by the derived class.")
```
@elaf if you could add the newlines you removed in this and the next file then its ok from my side

elafnm replied:
Greetings dear @philbucher and very good day to you,

kindly find updated changes, I used exactly the same files from the master branch,

thank you for your usual support we highly appreciate it, and kindly accept my apology for any inconvenience; honestly, this is my first experience with git and pull request,

best regards,

elafnm committed:
updated

elafnm committed:
required changes for ALE applications

pooyan-dadvand commented:
Just as a general recommendation, I would suggest you to take a look at the Codacy/PR Quality review.  It gives you automatic comments about the quality of your code.

In this PR you have more than 80 new issues which is far from optimum.

elafnm commented:
Greetings dear @pooyan-dadvand,

thank you for highlighting these concern, I was not aware of it. Sure will do, will try to work it out at earliest.

Thanks and best regards,

elafnm committed:
updates 1 for codacy

elafnm committed:
updates 2 for codacy

elafnm committed:
updates 3 for codacy

elafnm committed:
updating the variables names

elafnm committed:
updating the headers

elafnm committed:
updates 4 for codacy

elafnm committed:
updates 5 for codacy

elafnm committed:
updates 6 for codacy

elafnm committed:
updates 7 for codacy

elafnm commented:
Greetings dear @pooyan-dadvand,

kindly note that total number of issues were reduced to four,

also, for the remaining four issues, I tried to eliminate them but I could not, three of are mainly for 
"Variable is reassigned a value before the old one has been used." ,, where sometimes we need to initializing the matrix as zero before we start using it,,, or assign false value for boolean flag,

please let me know if you have further recommendations and I will be very pleased to do so,

thanks a lot and my best regards,

pooyan-dadvand commented:
Thanks for your effort to reduce number of issues. I think the first one (the `is_inside`) is avoidable but I understand that in other ones you are resizing the matrix. so you can leave them.

jcotela commented on applications/FluidDynamicsApplication/python_scripts/vms_monolithic_solver.py:
```diff
@@ -380,4 +380,4 @@ def CreateSolver(model_part, config):
             config.linear_solver_config)
 
-    return fluid_solver
+    return fluid_solver
```
Please revert this file

elafnm replied:
greetings dear @jcotela,

1. kindly note that today I downloaded kratos-master again, and then copied and pasted the vms_monolithic_solver file as is. I hope this will resolve this issue. If not please let me know and I will be pleasure to do any further modifications.

2. for the "documents/kratos_doxyfile", (I removed it by mistake when I run this file for reading it and then deleted it by mistake). Sorry about that, and I just added it today.

3. for the .bin file inside "tests/TwoFluidDarcyTest", I just removed it as well.

Thanks and best regards,

jcotela commented on documents/kratos_doxyfile:
```diff
@@ -1,1611 +0,0 @@
-# Doxyfile 1.6.3
```
Why was this file removed? We'd like to keep it.

jcotela reviewed (dismissed):
Dear @elafnm, 
We are getting there, but I still have to request a few more changes to this PR. 

Besides the comments, the file `applications/FluidDynamicsApplication/tests/TwoFluidDarcyTest/darcy_channel_test.post.bin` is listed as added by this PR. Please remove it.

elafnm committed:
Delete darcy_channel_test.post.bin

elafnm committed:
reverting vms-monolithic-solver and adding kratos-doxyfile

elafnm committed:
Merge branch 'master' of https://github.com/KratosMultiphysics/Kratos into HEAD

elafnm committed:
detached head

elafnm committed:
Merge branch 'b6' into elaf/ULF_surface_tension

elafnm committed:
changin head detached error

elafnm committed:
latest update, is_inlet flag

elafnm commented:
Greeting dear Dr. @pooyan-dadvand ,

thank you for your feedback, please note that I changed  `bool is_inlet=false` to  `bool is_inlet(false)` and I am not receiving any more issue about it; and the code is compiled fine and updated accordingly.

best regard,

jcotela committed:
Reverting endline in vms.h.

jcotela commented:
Dear @elafnm , I went  ahead and reverted the changes to vms.h since it was the only remaining "blocker" related to the fluid dynamics application. I'll be removing my negative review now, but please follow the advice of the other reviewers to go through with this PR.

elafnm commented:
Greetings dear @jcotela,

Thank you very much dear, we highly appreciate your effort and support,

best regards,

elafnm commented:
Greetings dear @philbucher and very good day to you,

Just would like kindly to know if I successfully fulfilled your valuable comments, and in case if you need me to do any further modification I will be very pleased to do so.

with our best regards,

elafnm commented:
Greetings dear @rubenzorrillar and very good day to you,

Just would like kindly to know if I successfully fulfilled your valuable comments, and in case if you need me to do any further modification I will be very pleased to do so.

with our best regards,

philbucher reviewed (dismissed):
Ok from my side

rubenzorrilla commented:
My requests were mainly related to the FluidDynamics and ALE apps. If you manage to fix them, it's ok from my side too. Finally it would be nice if you can implement Codacy suggestions.

elafnm commented:
Greetings dear @philbucher,

Thank you very much for your usual support and kind feedback,

with our best regards,

elafnm commented:
Greetings dear @rubenzorrilla,

Thank you for your appreciated feedback.

Beside, regarding Codacy suggestions, I am done with almost most of them, and what are remaining mostly for:

_"Variable is reassigned a value before the old one has been used."_ ,, where sometimes we need to initializing the matrix as zero before we start using it.

Thanks and best regards,

elafnm commented:
@playpaolo 

Greetings dear Dr. Pavel,

kindly find below the files added inside the ULFApplication (noting please that we followed the new kratos format and headers):

1.we added newly needed variables (we redefined the names to be comprehensive as advised) locally inside "ULF_application_variables.h" and "ULF_application_variables.cpp",

2. inside "custom_elements" we added the required element files as: "surface_tension.h" & "surface_tension.cpp"

3. inside "pyton_scripts": we added the required python solver  as: "SurfaceTension_monolithic_solver",

4. for the meshing applications, as advised by you dear earlie we created a new file as ""trigen_droplet_refine.h"... where it has the TriGenDropletModeler()

 5. we added the following files: "custom_processes/calculate_adhesion_force.h",  "custom_processes/calculate_curvature.h",  "custom_processes/find_triple_point.h", "custom_processes/calculate_contact_angle.h",  "custom_processes/calculate_nodal_length.h", "custom_processes/find_nodal_neighbours_surface_process.h", and  "custom_processes/calculate_normal_eq.h"

6. above files (in point#5) were added inside "add-process-to-python" as:
#include "custom_processes/calculate_adhesion_force.h"
include "custom_processes/calculate_curvature.h"
#include "custom_processes/find_triple_point.h"
#include "custom_processes/calculate_contact_angle.h"
#include "custom_processes/calculate_nodal_length.h"
#include "custom_processes/find_nodal_neighbours_surface_process.h"
#include "custom_processes/calculate_normal_eq.h"

 7. we made two test files as:
"lagrangian_droplet_test" and  "lagrangian_sessile_droplet_test" where they have only the required files ".mdpa" and "solvers"

thanks and best regards,

playpaolo reviewed (changes requested):
Dear @elafnm ,

please move the two tests you added inside the test_examples folder.

Moreover, after downloading your branch I tried executing your examples. The following error appears:

lagrangian_model_part.AddNodalSolutionStepVariable(VISCOUS_STRESSX)
RuntimeError: in void Kratos::VariablesList::Add(const Kratos::VariableData&) [ /home/pavel/kratos/kratos/kratos/containers/variables_list.h , Line 338 ]

with subject    :  Adding uninitialize variable to this variable list. Check if all variables are registered before kernel initialization 


Please make sure that you added all the necessary new variables inside the application.

Best,
Pavel

elafnm committed:
adding test_file

elafnm committed:
y
Xx
Merge branch 'elaf/ULF_surface_tension' of https://github.com/KratosMultiphysics/Kratos into elaf/ULF_surface_tension

x
X

elafnm committed:
added inside test_file

elafnm commented:
@playpaolo 

Greetings dear Dr. Pavel,

Thank you for your appreciated feedback. Accordingly, please find our latest updates as followed:

1. we just added the "test_example" folder, and moved our examples inside it.

2. all related variables, including "VISCOUS_STRESSX",  are defined locally (inside ULF_application_variables.h and ULF_application_variables.cpp). Also, to make sure nothing is missing, now we just merged and pulled our branch into my PC and compiled kratos again, and then run our examples successfully.

with our best regards,

playpaolo commented:
Hi @elafnm ,  I saw that the variables are declared in the variables file of the application.

By the way, you add this variable (VISCOUS_STRESSX) inside SurfaceTension_monolithic_solver.py. Why are you adding it here again (inside the lagrangian_drop_test.py)? Same holds for all the other variables you add to model part inside the solver. you do not need to add them again in the script of your example. 

Now I pulled the branch again and it does not compile. I will check out whats happening and write back to you.

elafnm committed:
removing duplicated variables add to model part inside the solvers

elafnm commented:
@playpaolo 

Greetings dear Dr.Pavel and very good day to you,

regarding the first point, _"Why are you adding it here again",_ sorry about that it is my bad I thought they need to be defined again. Now they are removed and updated inside the branch dear.

regarding compiling the code, today I tried one more time to pull our branch from different PC (note please that also I pulled the updated master branch), and they worked fine with us, and the examples were running smoothly, as you can kindly find in the below attachment please.

![ulf_built_target](https://user-images.githubusercontent.com/30217273/36768337-efb2cc80-1bfb-11e8-9508-800cdc5aa90f.jpeg)

![kratos_built](https://user-images.githubusercontent.com/30217273/36768365-13c4c812-1bfc-11e8-8e71-441867c25d38.jpeg)

![exampled run successfully](https://user-images.githubusercontent.com/30217273/36768284-ac14c9f6-1bfb-11e8-9afa-a79bd88c022c.jpeg)

playpaolo commented:
Hi @elafnm, just now I pulled several other branches and they all compile, while yours doent compile anymore. There seem to be some problem in: /home/pavel/Kratos/applications/CMakeLists.txt

See the error below:

CMake Error at applications/CMakeLists.txt:4 (SET):
  Syntax error in cmake code at

    /home/pavel/Kratos/applications/CMakeLists.txt:4

  when parsing string

    \

playpaolo commented:
@elafnm , yes, the problem is in CMakeList.txt. See the CMakeList.txt of any other branch (or master) and compare it with yours. 
This is yours:
[CMakeLists.txt](https://github.com/KratosMultiphysics/Kratos/files/1766138/CMakeLists.txt)

And this is master's file:
[CMakeLists.txt](https://github.com/KratosMultiphysics/Kratos/files/1766134/CMakeLists.txt)

elafnm committed:
updating CMakeLists.txt inside ULF Application

elafnm commented:
@playpaolo 

Greeting dear Dr. Pavel,

Thank you for your usual support and feedback,

1. for the main "CMakeList.txt" inside "kratos", they seems to be identical in the master and our branch

2. for "CMakeList.txt" inside "kratos/applications": I partially agree with you, as I believe that for the version you sent me it was not the most updated one, but now it is changed/updated in the master itself and it is updated in our branch accordingly. You may kindly dear find below links:

https://github.com/KratosMultiphysics/Kratos/blob/elaf/ULF_surface_tension/applications/CMakeLists.txt

https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CMakeLists.txt

3. for the CMakeList" inside "kratos/applications/ULFapplication", yes dear we missed ".cpp", so we just modified this line "${CMAKE_CURRENT_SOURCE_DIR}/ULF_application_variables" to be "${CMAKE_CURRENT_SOURCE_DIR}/ULF_application_variables.cpp",
kindly please dear note also that we just went through "CMakeList.txt." inside master; and compared our "CMakeList.txt." to the the existing CMakeList.txt. files inside  ULF_Application, Fluid_dynamic applications, and ALE_applications. and we hope it will work fine.

Thanks and best regards,

roigcarlo commented:
Hi, everything seems to work fine in my side. @playpaolo the problem with the backslash is probably in your local machine, Travis & Appveyor automatically check `kratos/CMakeLists.txt` and `kratos/applications/CMakeLists.txt` so if the checks pass like in a6d3d9a that means that the problem had to be in something not tested (for example, the missing `.cpp`).

playpaolo commented:
Hi @elafnm , finally we managed to compile your branch on my machine. I have run the tests and they seem to work fine. I have three comments: 1) please remove the empty Lagrangian_droplet_test and ...sessile test folders  from  ULFApplication/ . 2) You are importing quite many applications inside your examples. Do you really need all of them?  
from KratosMultiphysics import *
from KratosMultiphysics.IncompressibleFluidApplication import *
from KratosMultiphysics.FluidDynamicsApplication import *
from KratosMultiphysics.ExternalSolversApplication import *
from KratosMultiphysics.MeshingApplication import *
from KratosMultiphysics.ULFApplication import *
from KratosMultiphysics.StructuralApplication import *
#from KratosMultiphysics.PFEMApplication import *
from KratosMultiphysics.ALEApplication import *

Please leave only the necessary ones.

3) What should happen in the two examples? In the first one nothing seems to happen (circular droplet) remains static. In the second one (sessile) the tripple points move a bit, but very little. Initial configuration is close to the final one. I suggest you to add a readme.txt to each example explaining what you are modeling and what should one observe executing these examples.
Otherwise, from my side, you can commit your brunch.

playpaolo commented:
@roigcarlo Thanks Carlos. The problem is solved!

elafnm committed:
updated solvers, and added readme.txt files

elafnm commented:
@playpaolo 

Greetings dear Dr.Pavel and very good day to you,

Thank you for your approval, we highly appreciate it.

kindly please find our latest updates as:

1. regarding the empty Lagrangian_droplet_test and sessile test folders inside the ULFApplication, I just checked them and believe already removed; thought, please let me know if not so I will try to do so.

2. we removed the unused applications as advised,

3. we added two "readme.txt" files, one for each test_application,

please, let me know if you would like me to do any further modification, it will be my pleasure to do so.

with our best regards,

elafnm commented:
Greetings dear @rubenzorrilla and very good day,

Just would like kindly to inform you that we were working on fulfilling your valuable recommendations, and our work so far has been accepted by the team. Please let us know if you would like us to do any further modifications and it will be our pleasure to do so.

Thanks and best regards,

rubenzorrilla commented:
If you managed to address all the comments and suggestions, it is ok to me.

elafnm committed:
updated

elafnm commented:
Greetings dear @rubenzorrilla ,

Thank you for your appreciated acceptance,

with our best regards,

elafnm commented:
@pooyan-dadvand 

Greetings Dr. Pooyan and very good day to you dear,

It is our pleasure to inform you that with the support of Kratos team we could fulfill the requirements related to this pull request, and got the their appreciated acceptance.

thanks and best regards,

philbucher commented on applications/FluidDynamicsApplication/custom_strategies/strategies/residualbased_predictorcorrector_velocity_bossak_scheme_turbulent.h:
```diff
@@ -326,8 +326,21 @@ namespace Kratos {
-                        UpdateDisplacement(CurrentDisplacement, OldDisplacement, OldVelocity, OldAcceleration, CurrentAcceleration);
+    
+                        
+                        if((itNode)->FastGetSolutionStepValue(IS_LAGRANGIAN_INLET) == 0.0)
```
@elafnm I just saw that you are comparing a double with `==` 
this is dangerous due to machine accuracy, it might or might not work
you should do sth like the following:
`if((itNode)->FastGetSolutionStepValue(IS_LAGRANGIAN_INLET) < 1e-15)`

elafnm replied:
Thank you dear @philbucher  for this important observation,

I just changed it as recommended.

with our best regards,

philbucher replied:
Very good.
I think I saw it also in some other files, so you might want to check those too

elafnm replied:
@philbucher 

will do dear, and will keep it in mind for future as well.

with my best regards,

elafnm committed:
< 1e-15 instead of ==0

elafnm committed:
< 1e-15 instead of ==0 for IS_LAGRANGIAN_INLET

elafnm commented:
@philbucher 

done dear,

best regards,

pooyan-dadvand commented:
It seems that all are conformed with the PR but nobody is approving it. @playpaolo would you please approve it if you think it is ok?

@elafnm after approval you can merge the PR and please don't forget to remove the branch.(in this same page you can do both) 

If you want to work more on this part you may open another branch with a descriptive name of the specific feature you want to add.

elafnm commented:
Greetings dear @pooyan-dadvand and very good day to you,

thank you very much for your prompt and appreciate feedback and support,

please note I got these two emails around week ago:

"@ playpaolo approved this pull request".
and
"@ rubenzorrilla approved this pull request."

but yet the merging is blocked, with this message: "Merging can be performed automatically with 1 approving review."

thanks and best regards,

playpaolo commented:
I am approving it @pooyan-dadvand . Thanks!

jcotela commented:
@elafnm note that positive reviews are discarded when you push more commits to the branch. We only allow merging branches if the last version has been approved.
